### PR TITLE
Documentation for neural nets in MindMeld

### PR DESCRIPTION
**Pull request by murali1996**
This PR adds documentation to various neural nets models now available as part of MindMeld. Furthermore, few add-on code changes are also made part of this PR, for e.g. support for non-whitespace tokenizing bert-like models like distilroberta. (The minor code changes have been incorporated in this PR itself to avoid multiple PRs on the same topic.)

On the documentation front, the following changes are incorporated:
- a new page has been added called `deep_neural_nets.rst` that outlines the details of various deep neural models available as well as the params that are configurable for each of the model. 
- contents of previously existing `lstm.rst` page are now merged into the `deep_neural_nets.rst` page and the `lstm.rst` page has been deleted
- few reordering of pages has been done so the list is bit better aesthetically.

On the code changes front, the following changes are incorporated:
- correct few mistakes in code comments and improve error messages 
- add support for huggingface pretrained tokenizers that do not split text at whitespace (like roberta's tokenizer)
- distinctly address input encoding for sequence classification models vs. token classification models

### CODE REVIEW

File: mindmeld/models/nn_utils/classification.py [hunk 1]
```diff
@@ -221,6 +224,10 @@ def fit(self, examples: List[str], labels: Union[List[int], List[List[int]]], **
             self.train()
             optimizer.zero_grad()
             train_loss, train_batches = 0.0, 0.0
+            indices = list(range(len(train_examples)))
+            random.shuffle(indices)
+            train_examples = [train_examples[ii] for ii in indices]
```
Comment (murali1996) [inline mindmeld/models/nn_utils/classification.py:229]: Added shuffling of training data between epochs.

File: mindmeld/models/nn_utils/input_encoders.py [hunk 2]
```diff
@@ -223,6 +223,15 @@ def __init__(self, **kwargs):
...
 
+        # Considers not splitting query text at whitespace if this param is set to True; useful for
+        # languages with non-whitespace script e.g. Japanese, Chinese, etc.
+        # Cautions:
+        #   - when this param is set to True, any truncation of very long inputs due to the
+        #   tokenizer's max length is not checked for and can raise errors in token classification
+        #   - when this param is set to True within token classification, ensure that number of
+        #   tokens (without any distinction of whitespace) matches the number of entity tags/labels
+        self.avoid_whitespace_splitting = kwargs.get("avoid_whitespace_splitting")
+
```
Comment (murali1996) [inline mindmeld/models/nn_utils/input_encoders.py:234]: this new param is useful when working with non-whitespace scripted language

File: mindmeld/models/nn_utils/classification.py [hunk 3]
```diff
@@ -239,6 +246,7 @@ def fit(self, examples: List[str], labels: Union[List[int], List[List[int]]], **
                 })
                 batch_data = self.forward(batch_data)
                 loss = batch_data["loss"]
+                # .cpu() returns copy of tensor in CPU memory; also see https://tinyurl.com/3b4uvkj2
```
Comment (vembar) [inline mindmeld/models/nn_utils/classification.py:249]: stray?

File: source/userguide/deep_neural_nets.rst [hunk 4]
```diff
@@ -0,0 +1,1848 @@
+Deep Neural Networks in MindMeld
+================================
+
+The advent of `Deep learning <https://en.wikipedia.org/wiki/Deep_learning>`_ has led to substantial developments in the area of Conversational AI as well as Natural Langauge Processing (NLP) in general.
+Specifically, the proliferation of deep neural models based on :wiki_api:`Convolutional neural networks (CNN) <Convolutional_neural_network#Natural_language_processing>`, :wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` and :wiki_api:`Transformers <Transformer_(machine_learning_model)>`, and their rapid adaptation to wide range of NLP tasks has been a great success.
+Thus extending the suite of traditional machine learning models (e.g. :sk_guide:`Logistic regression <linear_model.html#logistic-regression>`, :sk_guide:`Decision tree <tree.html#tree>`, etc.), MindMeld provides the option for users to train and use a variety of deep neural models with an array of configurable parameters.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]:
```suggestion
MindMeld extends its suite of traditional machine learning models (e.g. :sk_guide:`Logistic regression <linear_model.html#logistic-regression>`, :sk_guide:`Decision tree <tree.html#tree>`, etc.) with a variety of deep neural models with an array of configurable parameters.
```

File: source/userguide/deep_neural_nets.rst [hunk 5]
```diff
@@ -0,0 +1,1848 @@
...
+
+The deep neural models are available both for text classification aka. sequence classification (:ref:`domain classification <domain_classification>` and :ref:`intent classification <intent_classification>`) as well as for token classification (:ref:`entity recognition <entity_recognition>`) tasks.
+The models are implemented using `Pytorch <https://pytorch.org/>`_ framework and thus need extra installations before starting to use them in your chatbot application.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]: Think you can move this to the note?

File: source/userguide/deep_neural_nets.rst [hunk 6]
```diff
@@ -0,0 +1,1848 @@
...
+
+.. note::
+
+   Please make sure to install the Pytorch requirement by running in the shell: :code:`pip install mindmeld[torch]`.
+
+MindMeld supports the use of pretrained transformer models such as :wiki_api:`BERT <BERT_(language_model)>` through the popular `Huggingface Transformers <https://huggingface.co/docs/transformers/index>`_ library.
+Several pretrained models from their `Models Hub <https://huggingface.co/models>`_ that can be used for sequence classification or token classification can be employed in your chatbot application.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]:
```suggestion
Several pretrained models from their `Models Hub <https://huggingface.co/models>`_ can be used for sequence or token classification to be employed in your chatbot application.
```

File: source/userguide/deep_neural_nets.rst [hunk 7]
```diff
@@ -0,0 +1,1848 @@
...
+
+.. note::
+
+   Install the extra *transformers* requirement by running in the shell: :code:`pip install mindmeld[transformers]`.
+
+Before proceeding to use the deep neural models, consider the following possible advantages and disadvantages of using them in place of traditional machine learning models.
+
+Some advantages are:
+
+- The deep models generally outperform traditional machine learning models on training sets with at least several hundreds or few thousands of queries when training from scratch, and with at least few hundreds if fine-tuning from a pretrained checkpoint.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]:
```suggestion
- The deep models generally outperform traditional machine learning models on training sets with several hundreds or few thousands of queries when training from scratch, and with at least few hundreds if fine-tuning from a pretrained checkpoint.
```

Comment (vembar) [reply source/userguide/deep_neural_nets.rst]: Actually, not sure this is an advantage or a disadvantage. Depends on the amount of data a user has.

File: source/userguide/deep_neural_nets.rst [hunk 8]
```diff
@@ -0,0 +1,1848 @@
...
+- On GPU-enabled devices, the deep networks can achieve training time comparable to some of the traditional models in MindMeld.
+
+Some disadvantages are:
+
+- Training time on CPU-only machines is a lot slower than for traditional machine learning models.
+- No automated hyperparameter tuning methods like :sk_api:`sklearn.model_selection.GridSearchCV <sklearn.model_selection.GridSearchCV.html>` are available for deep neural models.
+- The deep neural models generally occupy similar or more disk storage space compared to their traditional counterparts.
+
+Parameter tuning for deep neural models is more involved than for traditional machine learning models.
+A good starting point for understanding this subject is Andrej Karpathy's `course notes <https://cs231n.github.io/neural-networks-3/#baby>`_ from the Convolutional Neural Networks for Visual Recognition course at Stanford University.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]: Say something like - While we have good defaults for all params that works well across datasets, you can understand this better using this course.

File: source/userguide/deep_neural_nets.rst [hunk 9]
```diff
@@ -0,0 +1,1848 @@
...
+
+.. note::
+
+   To use deep neural networks instead of traditional machine learning models, simply make few modifications to the classifier configuration dictionaries for all or selected classifiers in your app's ``config.py``.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]: If we're not mentioning what these changes are or linking to a page that mentions them, this note seems unnecessary?

File: source/userguide/deep_neural_nets.rst [hunk 10]
```diff
@@ -0,0 +1,1848 @@
...
+
+Recall from :ref:`Working with the Domain Classifier <domain_classifier_configuration>` and :ref:`Working with the Intent Classifier <intent_classifier_configuration>` sections that a text classifier configuration consists of the keys
+
+- ``'features'``,
+- ``'param_selection'``,
+- ``'model_settings'``, and
+- ``'params'``
+
+amongst other keys that do not have distinction between traditional models or deep neural models.
+When working with deep neural models, the ``'features'`` and ``'param_selection'`` keys in the classifier configuration are redundant as we neither have to handcraft any feature sets for modeling nor there is an automated hyperparameter tuning.
```
Comment (vembar) [inline source/userguide/deep_neural_nets.rst]:
```suggestion
When working with deep neural models, the only relevant keys to be configured are ``'model_settings'`` and ``'params'``. We do not need the``'features'`` and ``'param_selection'`` keys in the classifier configuration we don't use handcrafted feature sets and no automated hyperparameter tuning.
```

File: source/userguide/deep_neural_nets.rst [hunk 11]
```diff
@@ -0,0 +1,1848 @@
...
+The capacity of an LSTM in maintaining the temporal information is generally dependant on its *hidden* dimension.
+Further, several LSTM layers can be stacked one-after-another and each layer can process the text from just beginning-to-end or both ways.
+The first layer's output embedding sequence is passed through the stacked LSTMs, which finally produces one vector per token of the input text.
+
+To obtain a single vector per word per input text, the vectors of all tokens corresponding to each word (for which an entity tag is to be ascertained) are pooled.
+This is unlike sequence classification models where all tokens of all words are pooled together.
+The pooled outputs are then passed through a classification layer.
+Dropout layers are used as regularizers to avoid over-fitting, which is a more common phenomenon when working with small sized datasets.
+
+Following are the different optional params that are configurable with the ``'lstm'`` classifier type.
```
Comment (berlinguist) [inline source/userguide/deep_neural_nets.rst:905]: `'lstm-pytorch'` classifier type?

File: source/userguide/deep_neural_nets.rst [hunk 12]
```diff
@@ -0,0 +1,1848 @@
...
```
Comment (berlinguist) [inline source/userguide/deep_neural_nets.rst:1052]: `'cnn-lstm'` classifier type?

File: source/userguide/deep_neural_nets.rst [hunk 13]
```diff
@@ -0,0 +1,1848 @@
...
+   }
+
+4. ``'lstm-lstm'`` classifier type
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _dnns_token_classification_models_lstm_lstm:
+
+:wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` based text classifiers are classifiers that utilize recurrent feedback connections to be able to learn temporal dependencies in sequential data.
+When coupled with :wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` for extracting character-level features from input text, the overall architecture can better model the textual data as well as it is more robust to variations in the spellings.
+In its typical architecture for entity recognition, the first layer embeds the sequence of words obtained f
```
Comment (berlinguist) [inline source/userguide/deep_neural_nets.rst:1236]: `'lstm-lstm'` classifier type?

Comment (berlinguist) [inline source/userguide/deep_neural_nets.rst:1430]: This section is structured differently than the other 4 sections of this type, in that it has a straightforward bulleted list of the key points for this section. The other sections do a good job of laying out the key points, too, but they do so in paragraph form. 

I'm kind of partial to this display of information, though it is a bit of a departure from what appears to be the norm for MindMeld documentation. On the one hand, it might be a lot of work to rewrite the other sections in this 'Entity recognition'  section as opposed to making this conform to the style imposed in those other sections. But on the other hand, this is a really nice point-for-point outline of what is happening with this classifier type, and it's presented in a very user-friendly visual and organizational style.

**Review by berlinguist — COMMENTED** [review]
A few questions & a comment about structuring sections in this documentation

File: source/userguide/deep_neural_nets.rst (hunk 4 above)
Comment (murali1996) [reply source/userguide/deep_neural_nets.rst]: Related changes made in [Kelsey's PR](https://github.com/cisco/mindmeld/blob/feature/doc-review-nn_support_phase2_mm_docs/source/userguide/deep_neural_nets.rst)

File: source/userguide/deep_neural_nets.rst [hunk 14]
```diff
@@ -0,0 +1,1848 @@
...
+Specifically, the proliferation of deep neural models based on :wiki_api:`Convolutional neural networks (CNN) <Convolutional_neural_network#Natural_language_processing>`, :wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` and :wiki_api:`Transformers <Transformer_(machine_learning_model)>`, and their rapid adaptation to wide range of NLP tasks has been a great success.
+Thus extending the suite of traditional machine learning models (e.g. :sk_guide:`Logistic regression <linear_model.html#logistic-regression>`, :sk_guide:`Decision tree <tree.html#tree>`, etc.), MindMeld provides the option for users to train and use a variety of deep neural models with an array of configurable parameters.
+
+The deep neural models are available both for text classification aka. sequence classification (:ref:`domain classification <domain_classification>` and :ref:`intent classification <intent_classification>`) as well as for token classification (:ref:`entity recognition <entity_recognition>`) tasks.
```
[the remark this answers is not in the document]
Comment (murali1996) [reply source/userguide/deep_neural_nets.rst]: Related changes made in [Kelsey's PR](https://github.com/cisco/mindmeld/blob/feature/doc-review-nn_support_phase2_mm_docs/source/userguide/deep_neural_nets.rst)

File: source/userguide/deep_neural_nets.rst (hunk 5 above)
Comment (murali1996) [reply source/userguide/deep_neural_nets.rst]: Related changes made in [Kelsey's PR](https://github.com/cisco/mindmeld/blob/feature/doc-review-nn_support_phase2_mm_docs/source/userguide/deep_neural_nets.rst)

**Review by berlinguist — APPROVED** [review]
Looks great :)

File: mindmeld/models/nn_utils/helpers.py [hunk 15]
```diff
@@ -187,10 +192,9 @@ class TokenClassificationType(enum.Enum):
         "output_keep_prob": 1.0,  # this dropout unnecessary upon using `embedder_output_keep_prob`
         # other values for following keys will possibly throw errors
         "embedder_type": "bert",
-        "add_terminals": True,
         "tokenizer_type": TokenizerType.HUGGINGFACE_PRETRAINED_TOKENIZER.value,
         # keys that are not mutually exclusive and are valid when some of the above keys are set
-        "save_frozen_bert_weights": False,  # the key is valid only when update_embeddings=True
+        "save_frozen_embedder": False,  # the key is valid only when update_embeddings=False
```
Comment (vembar) [inline mindmeld/models/nn_utils/helpers.py:197]: Seems like this key is  just the opposite of `update_embeddings`?

**Review by vembar — APPROVED** [review]
Couple of questions that may need to be resolved, but otherwise looks good.

**Comment by minhtuev** [comment]
Exciting stuffs!
